odoo/tools/safe_eval.py

```diff
@@ -1,225 +1,576 @@
 
-    # set operations are almost twice as fast as a manual iteration + condition
-    # when loading /web according to line_profiler
-    code_codes = {i.opcode for i in dis.get_instructions(code_obj)}
-    if not allowed_codes >= code_codes:
-        raise ValueError("forbidden opcode(s) in %r: %s" % (expr, ', '.join(opname[x] for x in (code_codes - allowed_codes))))
+    __slots__ = ["__obj", "__type_checker"]
 
-    for const in code_obj.co_consts:
-        if isinstance(const, CodeType):
-            assert_valid_codeobj(allowed_codes, const, 'lambda')
+    def __init__(self, obj, type_checker):
+        if isinstance(obj, _wrapped_instances) or (type(obj) is type and obj in _wrapped_types) or type(obj) is types.ModuleType and obj.__name__ in _ALLOWED_MODULES:  # noqa: E721 (Reason: `type(...) is ...` is stricter than `isinstance(..., ...)` in this case) # pylint: disable=unidiomatic-typecheck
+            self.__obj = obj
+        else:
+            self.__obj = type_checker(obj)
+
+        self.__type_checker = type_checker
+
+    def __getitem__(self, item):
+        self.__type_checker(item)
+        return self.__type_checker(self.__obj.__getitem__(item))
+
+    def __setitem__(self, key, value):
+        self.__type_checker(key)
+        self.__type_checker(value)
+
+        if isinstance(key, str) and "__" in key:
+            raise NameError(f"Forbidden name '{key}'.")
+
+        self.__obj.__setitem__(key, value)
+
+    def __call__(self, *args, **kwargs):
+        for arg in (*args, *kwargs.values()):
+            self.__type_checker(arg)
+
+        return self.__type_checker(self.__obj(*args, **kwargs))
+
+    def __getattr__(self, name):
+        if "__" in name and not name.endswith("__"):
+            raise NameError(f"Forbidden name '{name}'.")
+
+        obj = super().__getattribute__("_SafeWrapper__obj")
+
+        if (isinstance(obj, str) or (type(obj) is type and issubclass(obj, str))) and name in ("format", "format_map"):
+            raise SyntaxError("format method is forbidden on strings.")
+
+
+        return self.__type_checker(getattr(obj, name))
+
+
+# Set of the types that are allowed to be used in eval
+# Almost all of them are builtin types
+_ALLOWED_TYPES = frozenset({
+    bool,
+    bytes,
+    complex,
+    datetime.date,
+    datetime.datetime,
+    datetime.time,
+    datetime.timedelta,
+    dict,
+    enumerate,
+    filter,
+    float,
+    int,
+    JSON,
+    list,
+    map,
+    odoo.exceptions.UserError,
+    range,
+    relativedelta,
+    reversed,
+    set,
+    slice,
+    str,
+    tuple,
+    zip,
+})
 
-def test_expr(expr, allowed_codes, mode="eval", filename=None):
-    """test_expr(expression, allowed_codes[, mode[, filename]]) -> code_object
+# Set of the instances that are allowed
+# This means that the value of their __class__ attributes are denied,
+# Which means that they can't be re-created
+_ALLOWED_INSTANCES = frozenset({
+    EmailMessage,
+    EnumMeta,
```

Comment:
After a lot of research, There is [this class](https://github.com/odoo/odoo/blob/8e9509d8a492358908c78c65a8e18900becaf6f4/odoo/fields.py#L4023) that we use a lot staticly and its type is `EnumMeta` before Python3.11 and after 3.11 it will be changed to `EnumType`

Reply:
why not just adding Command ?